In [3]:
import pandas as pd
import polars as pl
import numpy as np

# **Polars**

1. Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polar

In [ ]:
df = pl.read_csv('train.csv')

2. Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д. - 1 балл
hint: используйте методы и атрибуты датафрейма: describe(), dtypes, null_count(), mean()...

In [ ]:
type(df)

polars.dataframe.frame.DataFrame

In [ ]:
df.shape

(891, 12)

In [ ]:
df.schema

Schema([('PassengerId', Int64),
        ('Survived', Int64),
        ('Pclass', Int64),
        ('Name', String),
        ('Sex', String),
        ('Age', Float64),
        ('SibSp', Int64),
        ('Parch', Int64),
        ('Ticket', String),
        ('Fare', Float64),
        ('Cabin', String),
        ('Embarked', String)])

In [ ]:
df.dtypes

[Int64,
 Int64,
 Int64,
 String,
 String,
 Float64,
 Int64,
 Int64,
 String,
 Float64,
 String,
 String]

In [ ]:
df.glimpse()

Rows: 891
Columns: 12
$ PassengerId <i64> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
$ Survived    <i64> 0, 1, 1, 1, 0, 0, 0, 0, 1, 1
$ Pclass      <i64> 3, 1, 3, 1, 3, 3, 1, 3, 3, 2
$ Name        <str> 'Braund, Mr. Owen Harris', 'Cumings, Mrs. John Bradley (Florence Briggs Thayer)', 'Heikkinen, Miss. Laina', 'Futrelle, Mrs. Jacques Heath (Lily May Peel)', 'Allen, Mr. William Henry', 'Moran, Mr. James', 'McCarthy, Mr. Timothy J', 'Palsson, Master. Gosta Leonard', 'Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)', 'Nasser, Mrs. Nicholas (Adele Achem)'
$ Sex         <str> 'male', 'female', 'female', 'female', 'male', 'male', 'male', 'male', 'female', 'female'
$ Age         <f64> 22.0, 38.0, 26.0, 35.0, 35.0, null, 54.0, 2.0, 27.0, 14.0
$ SibSp       <i64> 1, 1, 0, 1, 0, 0, 0, 3, 0, 1
$ Parch       <i64> 0, 0, 0, 0, 0, 0, 0, 1, 2, 0
$ Ticket      <str> 'A/5 21171', 'PC 17599', 'STON/O2. 3101282', '113803', '373450', '330877', '17463', '349909', '347742', '237736'
$ Fare        <f64> 7.25, 71.2833, 7

In [ ]:
df.describe()

statistic,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
str,f64,f64,f64,str,str,f64,f64,f64,str,f64,str,str
"""count""",891.0,891.0,891.0,"""891""","""891""",714.0,891.0,891.0,"""891""",891.0,"""204""","""889"""
"""null_count""",0.0,0.0,0.0,"""0""","""0""",177.0,0.0,0.0,"""0""",0.0,"""687""","""2"""
"""mean""",446.0,0.383838,2.308642,null,null,29.699118,0.523008,0.381594,null,32.204208,null,null
"""std""",257.353842,0.486592,0.836071,null,null,14.526497,1.102743,0.806057,null,49.693429,null,null
"""min""",1.0,0.0,1.0,"""Abbing, Mr. Anthony""","""female""",0.42,0.0,0.0,"""110152""",0.0,"""A10""","""C"""
"""25%""",224.0,0.0,2.0,null,null,20.0,0.0,0.0,null,7.925,null,null
"""50%""",446.0,0.0,3.0,null,null,28.0,0.0,0.0,null,14.4542,null,null
"""75%""",669.0,1.0,3.0,null,null,38.0,1.0,0.0,null,31.0,null,null
"""max""",891.0,1.0,3.0,"""van Melkebeke, Mr. Philemon""","""male""",80.0,8.0,6.0,"""WE/P 5735""",512.3292,"""T""","""S"""


In [ ]:
df.null_count()

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32
0,0,0,0,0,177,0,0,0,0,687,2


In [ ]:
df.mean()

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
f64,f64,f64,str,str,f64,f64,f64,str,f64,str,str
446.0,0.383838,2.308642,null,null,29.699118,0.523008,0.381594,null,32.204208,null,null


3. Посчитайте количество пассажиров каждого класса (Pclass) - 1 балл
hint: используйте методы: get_column(), value_counts()...

In [ ]:
df.columns

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

In [ ]:
df.get_column('Pclass').value_counts().sort('Pclass')

Pclass,count
i64,u32
1,216
2,184
3,491


4. Выведите количество выживших мужчин и женщин на корабле - 1 балл
hint: используйте методы: groupby(), agg(), pl.col(), sum()...


In [ ]:
df.head(2)

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
i64,i64,i64,str,str,f64,i64,i64,str,f64,str,str
1,0,3,"""Braund, Mr. Owen Harris""","""male""",22.0,1,0,"""A/5 21171""",7.25,null,"""S"""
2,1,1,"""Cumings, Mrs. John Bradley (Fl…","""female""",38.0,1,0,"""PC 17599""",71.2833,"""C85""","""C"""


In [ ]:
df.group_by('Sex').agg(pl.col('Survived').sum())

Sex,Survived
str,i64
"""male""",109
"""female""",233


5. Выведите часть таблицы с пассажирами, возраст которых больше 44 лет - 1 балл
hint: используйте методы: filter(), pl.col()...

In [ ]:
df.filter(pl.col('Age') > 44)

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
i64,i64,i64,str,str,f64,i64,i64,str,f64,str,str
7,0,1,"""McCarthy, Mr. Timothy J""","""male""",54.0,0,0,"""17463""",51.8625,"""E46""","""S"""
12,1,1,"""Bonnell, Miss. Elizabeth""","""female""",58.0,0,0,"""113783""",26.55,"""C103""","""S"""
16,1,2,"""Hewlett, Mrs. (Mary D Kingcome…","""female""",55.0,0,0,"""248706""",16.0,null,"""S"""
34,0,2,"""Wheadon, Mr. Edward H""","""male""",66.0,0,0,"""C.A. 24579""",10.5,null,"""S"""
53,1,1,"""Harper, Mrs. Henry Sleeper (My…","""female""",49.0,1,0,"""PC 17572""",76.7292,"""D33""","""C"""
…,…,…,…,…,…,…,…,…,…,…,…
858,1,1,"""Daly, Mr. Peter Denis ""","""male""",51.0,0,0,"""113055""",26.55,"""E17""","""S"""
863,1,1,"""Swift, Mrs. Frederick Joel (Ma…","""female""",48.0,0,0,"""17466""",25.9292,"""D17""","""S"""
872,1,1,"""Beckwith, Mrs. Richard Leonard…","""female""",47.0,1,1,"""11751""",52.5542,"""D35""","""S"""


## **Ускорение работы с pandas**

1. Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas

In [ ]:
df_pd = pd.read_csv('train.csv')

In [ ]:
df_pd.head(2)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C


2. Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck - 1 бал
hint: используйте методы: bn.nanmean(), bn.nanstd()

In [ ]:
!pip install bottleneck
import bottleneck as bn

In [ ]:
%%timeit
mean_age_pl = df.select(pl.col('Age').mean())
mean_age_pl

28.7 µs ± 692 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
%%timeit
mean_age_pd = df_pd['Age'].mean()
mean_age_pd

33.5 µs ± 4.56 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
%%timeit
bn.nanmean(df_pd['Age'])

15.6 µs ± 3.08 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [ ]:
bn.nanmean(df_pd['Age'])

29.69911764705882

In [ ]:
bn.nanstd(df_pd['Age'])

14.516321150817317

3. Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new - 1 балл
hint: используйте методы: .itertuples() или apply()

In [ ]:
df_pd['Fare_new'] = df_pd['Fare'].apply(lambda x: x*1.3)
df_pd.head(2)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Fare_new
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,9.42500
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,92.66829


# **Оптимизация типов pandas**

1. Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [10]:
 housing = pd.read_csv('Housing.csv')
 housing[:3]

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished


2. Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях - 2 балла

In [11]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [12]:
housing.describe()

,price,area,bedrooms,bathrooms,stories,parking
count,5.450000e+02,545.000000,545.000000,545.000000,545.000000,545.000000
mean,4.766729e+06,5150.541284,2.965138,1.286239,1.805505,0.693578
std,1.870440e+06,2170.141023,0.738064,0.502470,0.867492,0.861586
min,1.750000e+06,1650.000000,1.000000,1.000000,1.000000,0.000000
25%,3.430000e+06,3600.000000,2.000000,1.000000,1.000000,0.000000
50%,4.340000e+06,4600.000000,3.000000,1.000000,2.000000,0.000000
75%,5.740000e+06,6360.000000,3.000000,2.000000,2.000000,1.000000
max,1.330000e+07,16200.000000,6.000000,4.000000,4.000000,3.000000


**числовые типы**
- Столбец Price - max = 1.330000e+07 - входит в диапазон int32
- столбец area  - max = 16200 - диапазон int16
- столбцы bedrooms/bathrooms/stories/parking  = 6 - диапазон int8

**тип object **

столбцы c YES|NO guestroom/basement/hotwaterheating/airconditioning/prefarea переводим в category

3. Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации - 1 балл

In [13]:
housing = housing.fillna(0)

In [14]:
housing['price'] = housing['price'].astype(np.int32)
housing['area'] = housing['area'].astype(np.int16)
housing['bedrooms'] = housing['bedrooms'].astype(np.int8)
housing['bathrooms'] = housing['bathrooms'].astype(np.int8)
housing['stories'] = housing['stories'].astype(np.int8)
housing['parking'] = housing['parking'].astype(np.int8)

In [15]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int32 
 1   area              545 non-null    int16 
 2   bedrooms          545 non-null    int8  
 3   bathrooms         545 non-null    int8  
 4   stories           545 non-null    int8  
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int8  
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int16(1), int32(1), int8(4), object(7)
memory usage: 35.3+ KB


In [16]:
housing['mainroad'].info(memory_usage="deep")

<class 'pandas.core.series.Series'>
RangeIndex: 545 entries, 0 to 544
Series name: mainroad
Non-Null Count  Dtype 
--------------  ----- 
545 non-null    object
dtypes: object(1)
memory usage: 27.7 KB


In [17]:
housing['mainroad'] = housing['mainroad'].astype("category")

In [18]:
housing['mainroad'].info(memory_usage="deep")

<class 'pandas.core.series.Series'>
RangeIndex: 545 entries, 0 to 544
Series name: mainroad
Non-Null Count  Dtype   
--------------  -----   
545 non-null    category
dtypes: category(1)
memory usage: 888.0 bytes


In [19]:
housing['mainroad'].unique()

['yes', 'no']
Categories (2, object): ['no', 'yes']

In [20]:
housing[:2]

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished


In [21]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int16   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    object  
 7   basement          545 non-null    object  
 8   hotwaterheating   545 non-null    object  
 9   airconditioning   545 non-null    object  
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    object  
 12  furnishingstatus  545 non-null    object  
dtypes: category(1), int16(1), int32(1), int8(4), object(6)
memory usage: 31.7+ KB


In [22]:
housing['guestroom'] = housing['guestroom'].astype("category")
housing['basement'] = housing['basement'].astype("category")
housing['hotwaterheating'] = housing['hotwaterheating'].astype("category")
housing['airconditioning'] = housing['airconditioning'].astype("category")

In [23]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int16   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    object  
 12  furnishingstatus  545 non-null    object  
dtypes: category(5), int16(1), int32(1), int8(4), object(2)
memory usage: 17.2+ KB


Выводы:
1) сначала оптимизируем int = числовые типы были уменьшены (все, кроме price) - съекономлено 36% памяти (55,5 КB / 35,3 KB)  
2) оптимизируем object = приведем к категориальным значениям
   - протестировала изменения по столбцу mainroad. До изменения типа memory usage: 27.7 KB, после memory usage: 888.0 bytes
   - перевела столбцы c YES|NO guestroom, basement, hotwaterheating, airconditioning и prefarea в category - общая экономия 69% (55,5 КB / 17,2 КB)

ИТОГО: получена значительная экономия памяти